spec.bs

```diff
@@ -70,11 +83,161 @@ dictionary AuctionAdInterestGroup {
+    1. [=map/for each=] |key| → |value| of |group|["{{AuctionAdInterestGroup/prioritySignalsOverrides}}"]:
+      1. Set |out|[|key|] to |value|.
+    1. Set |interestGroup|'s [=interest group/priority signals overrides=] to |out|.
+  1. Set |interestGroup|'s [=interest group/execution mode=] to
+    |group|["{{AuctionAdInterestGroup/executionMode}}"].
+  1. For each |groupMember| and |interestGroupField| in the following table <table class="data">
+      <thead><tr><th>Group member</th><th>Interest group field</th></tr></thead>
+      <tr>
+        <td>"{{AuctionAdInterestGroup/biddingLogicUrl}}"</td>
+        <td>[=interest group/bidding url=]</td>
+      </tr>
+      <tr>
+        <td>"{{AuctionAdInterestGroup/biddingWasmHelperUrl}}"</td>
+        <td>[=interest group/bidding wasm helper url=]</td>
+      </tr>
+      <tr>
+        <td>"{{AuctionAdInterestGroup/dailyUpdateUrl}}"</td>
+        <td>[=interest group/daily update url=]</td>
+      </tr>
+      <tr>
+        <td>"{{AuctionAdInterestGroup/trustedBiddingSignalsUrl}}"</td>
+        <td>[=interest group/trusted bidding signals url=]</td>
+      </tr>
+    </table>
+    1. If |group| [=map/contains=] |groupMember|:
+      1. Let |parsedUrl| be the result of running the [=URL parser=] on |group|[|groupMember|].
+      1. If |parsedUrl| is an error, return [=a promise rejected with=] a {{TypeError}}.
+      1. Set |interestGroup|'s |interestGroupField| to |parsedUrl|.
+  1. [=list/For each=] |key| of |group|["{{AuctionAdInterestGroup/trustedBiddingSignalsKeys}}"]:
+    1. [=list/Append=] |key| to |interestGroup|'s [=interest group/trusted bidding signals keys=].
+  1. If |group|["{{AuctionAdInterestGroup/userBiddingSignals}}"] [=map/exists=]:
+    1. Let |interestGroup|'s [=interest group/user bidding signals=] be the result of
+      [=serializing a JavaScript value to a JSON string=], given
+      |group|["{{AuctionAdInterestGroup/userBiddingSignals}}"]. If it fails, then return
+      [=a promise rejected with=] a {{TypeError}}.
+  1. For each |groupMember| and |interestGroupField| in the following table <table class="data">
+      <thead><tr><th>Group member</th><th>Interest group field</th></tr></thead>
+      <tr>
+        <td>"{{AuctionAdInterestGroup/ads}}"</td>
+        <td>[=interest group/ads=]</td>
+      </tr>
+      <tr>
+        <td>"{{AuctionAdInterestGroup/adComponents}}"</td>
+        <td>[=interest group/ad components=]</td>
+      </tr>
+    </table>
+    1. [=list/For each=] |ad| of |group|[|groupMember|]:
+      1. Let |igAd| be a new [=interest group ad=].
+      1. Let |renderUrl| be the result of running the [=URL parser=] on
+        |ad|["{{AuctionAd/renderUrl}}"].
+      1. If |renderUrl| is an error, return [=a promise rejected with=] a {{TypeError}}.
+      1. set |igAd|'s [=interest group ad/render url=] to |renderUrl|.
+      1. If |ad|["{{AuctionAd/metadata}}"] is not null, let |igAd|'s [=interest group ad/metadata=] 

```

Comment:
Is just `null` special, or also `undefined`? Are there any other disallowed types, from https://webidl.spec.whatwg.org/#idl-types?

Reply:
1. For |ad|, metadata field is optional, so it can be {renderUrl: "xxxx"}, without metadata field.
What I wanted to check was whether the javascript's |ad| object has metadata field, so I think my check about null is wrong. It should be:  If |ad|["{{AuctionAd/metadata}}"] [=map/exists=].
2. null is not special. The requirement is that the metadata must be JSON-serializable. The method will try to jsonify(metadata) and throw a TypeError if it fails.